File: jupyter2/answers/minc/minc/go/minc/main.go
package main

import (
	"fmt"
	"io"
	"os"
)

func readInput() (string, error) {
	if len(os.Args) > 1 {
		b, err := os.ReadFile(os.Args[1])
		return string(b), err
	}
	b, err := io.ReadAll(os.Stdin)
	return string(b), err
}

func writeOutput(asm string) error {
	if len(os.Args) > 2 {
		return os.WriteFile(os.Args[2], []byte(asm), 0644)
	}
	_, err := io.WriteString(os.Stdout, asm)
	return err
}

func fatal(err error) {
	fmt.Fprintln(os.Stderr, err)
	os.Exit(1)
}

func main() {
	src, err := readInput()
	if err != nil {
		fatal(err)
	}
	prog, err := ParseString(src)
	if err != nil {
		fatal(err)
	}
	asm, err := GenProgram(prog)
	if err != nil {
		fatal(err)
	}
	if err := writeOutput(asm); err != nil {
		fatal(err)
	}
}

File: jupyter2/answers/minc/minc/rs/minc/src/main.rs
mod ast;
mod codegen;
mod lex;
mod parse;

use std::fs;
use std::io::{self, Read, Write};
use std::process::exit;

fn run() -> Result<(), String> {
    let args: Vec<String> = std::env::args().collect();

    let src = if args.len() > 1 {
        fs::read_to_string(&args[1]).map_err(|e| e.to_string())?
    } else {
        let mut s = String::new();
        io::stdin().read_to_string(&mut s).map_err(|e| e.to_string())?;
        s
    };

    let prog = parse::parse_string(&src)?;
    let asm = codegen::gen_program(&prog)?;

    if args.len() > 2 {
        let mut f = fs::File::create(&args[2]).map_err(|e| e.to_string())?;
        f.write_all(asm.as_bytes()).map_err(|e| e.to_string())?;
    } else {
        io::stdout()
            .write_all(asm.as_bytes())
            .map_err(|e| e.to_string())?;
    }
    Ok(())
}

fn main() {
    if let Err(e) = run() {
        eprintln!("{}", e);
        exit(1);
    }
}
